### PR TITLE
TRUNK-5087 Added tests for Openmrsutil.java.

### PR DESCRIPTION
**Pull request by Andu033**
-Added tests for classes untested on coveralls .
-Solved a TODO for conceptListHelper 
-Rewrite conceptSetHelper because was all duplicate code.

### CODE REVIEW

**Comment by Andu033** [comment]
Used TDD to rewrite conceptSetHelper, so i wrote the tests verified that pass and then i rewrote the function , so everything works good.

**Comment by samuelmale** [comment]
Can you add the link to ticket?

**Comment by Andu033** [comment]
@samuelmale  of course.
Here: https://issues.openmrs.org/browse/TRUNK-5087?filter=-1

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 1]
```diff
@@ -19,34 +19,21 @@
...
 import org.apache.commons.io.IOUtils;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
-import org.openmrs.GlobalProperty;
-import org.openmrs.PatientIdentifier;
-import org.openmrs.PatientIdentifierType;
-import org.openmrs.User;
+import org.openmrs.*;
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Why use wildcard imports https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotusewildcardimports

File: api/src/main/java/org/openmrs/util/OpenmrsUtil.java [hunk 2]
```diff
@@ -715,37 +715,8 @@ public static int compareWithNullAsLatest(Date d1, Date d2) {
...
-			if (c != null) {
-				if (isSet) {
-					List<Concept> inSet = cs.getConceptsByConceptSet(c);
-					ret.addAll(inSet);
-				} else {
-					ret.add(c);
-				}
-			}
-		}
+		Set<Concept> ret = new HashSet<>(conceptListHelper(descriptor));
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/util/OpenmrsUtil.java]: Why did you remove the above code for this method and instead delegated `conceptListHelper(arg..)`; do you mean these two methods treat these collections in the same way?

File: api/src/main/java/org/openmrs/util/OpenmrsUtil.java [hunk 3]
```diff
@@ -812,16 +783,22 @@ public static int compareWithNullAsLatest(Date d1, Date d2) {
...
 		}
 		ConceptService cs = Context.getConceptService();
 		
 		for (StringTokenizer st = new StringTokenizer(descriptor, "|"); st.hasMoreTokens();) {
 			String s = st.nextToken().trim();
+			if(duplicates.contains(s)) {
+				continue;
+			}
+			else {
+			duplicates.add(s);
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/util/OpenmrsUtil.java]: * Can you correct this indention?
* By the way, why do we need need `duplicates`, why do we even need to cache them?

**Comment by samuelmale** [comment]
You also need to squash your comments into one.

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 1 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: You are right. My IDE did that . i will fix

File: api/src/main/java/org/openmrs/util/OpenmrsUtil.java (hunk 3 above)
Comment (Andu033) [reply api/src/main/java/org/openmrs/util/OpenmrsUtil.java]: i cached to solve that todo . The function didnt handle duplicates in descriptor .

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 4]
```diff
@@ -666,9 +674,185 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
 	/**
-	 * @see OpenmrsUtil#nullSafeEqualsIgnoreCase(String,String)
+	 * @see OpenmrsUtil#folderContains(File, String)
+	 */
+	@Test
+	public void folderContains_shouldReturnTrueWhenFolderContainsFileWithGivenName() {
+		String path = System.getProperty("user.dir");
+		File folder = new File(path+"/folder");
+		folder.mkdir();
+		File file = new File(path+"/folder/file");
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Will this file path work on all OSs! Will it work properly on windows? Can you use Java's abstraction and get system path symbols ie "/" or "\"

Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: You are right . I will change with System file separator now.

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 5]
```diff
@@ -666,9 +676,189 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+	
+	/**
+	 * @see OpenmrsUtil#folderContains(File, String)
+	 */
+	@Test
+	public void folderContains_shouldReturnFalseWhenGivenStringIsNull() {
+		String path = System.getProperty("user.dir");
+		String separator = FileSystems.getDefault().getSeparator();
+		File folder = new File(path + separator + "folder");
+		folder.mkdir();
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Don't you think your doing alot of code duplication of the above four lines of code in many methods?

Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Want me to use a private method?

Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Clean code should be reusable instead of 'copy & paste'. https://www.quora.com/How-to-reuse-a-code-in-Java

You may want to separate concerns by creating another method

File: api/src/main/java/org/openmrs/util/OpenmrsUtil.java [hunk 6]
```diff
@@ -715,37 +715,8 @@ public static int compareWithNullAsLatest(Date d1, Date d2) {
...
-			}
-			if (c != null) {
-				if (isSet) {
-					List<Concept> inSet = cs.getConceptsByConceptSet(c);
-					ret.addAll(inSet);
-				} else {
-					ret.add(c);
-				}
-			}
-		}
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/util/OpenmrsUtil.java]: I think the above change is beyond the scope for this ticket.

File: api/src/main/java/org/openmrs/util/OpenmrsUtil.java [hunk 7]
```diff
@@ -812,16 +783,22 @@ public static int compareWithNullAsLatest(Date d1, Date d2) {
 		return ret;
 	}
 	
-	// TODO: properly handle duplicates
 	public static List<Concept> conceptListHelper(String descriptor) {
 		List<Concept> ret = new ArrayList<>();
+		Set<String> duplicates = new HashSet<>();
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/util/OpenmrsUtil.java]: I believe implementing the TODO is also beyond the scope of this issue

Comment (Andu033) [reply api/src/main/java/org/openmrs/util/OpenmrsUtil.java]: Ok. I will revert

File: api/src/main/java/org/openmrs/util/OpenmrsUtil.java (hunk 6 above)
Comment (Andu033) [reply api/src/main/java/org/openmrs/util/OpenmrsUtil.java]: Ok. I will revert

**Comment by Andu033** [comment]
@samuelmale  i did the changes. What is left to do ?

**Comment by Andu033** [comment]
@samuelmale i updated the method name and the exception from catch block. The links was very useful . Is there anything left to adjust?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 8]
```diff
@@ -666,9 +683,169 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+		File folder = createFolderInUserDirectory("folder");
+		String separator = FileSystems.getDefault().getSeparator();
+		File file = new File(folder.getPath()  + separator + "file");
+		try {
+			file.createNewFile();
+			assertTrue(OpenmrsUtil.folderContains(folder, "file"));
+			file.delete();
+
+		} 
+		catch (IOException e) {}
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: What happens if the IOException is caught here? 
Does that mean we shall have no test assertions?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 9]
```diff
@@ -666,9 +683,169 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+			PrintWriter writer = new PrintWriter(file);
+			String output = "The first line" + "\n" + "The second line";
+			writer.println("The first line");
+			writer.print("The second line");
+			writer.close();
+			assertTrue(output.equals(OpenmrsUtil.getFileAsString(file)));
+			file.delete();
+			
+		}
+		catch (IOException e) {}
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: What happens if the IOException is caught here?
Does that mean we shall have no test assertions?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 10]
```diff
@@ -666,9 +683,169 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
 	}
 	
 	/**
-	 * @see OpenmrsUtil#nullSafeEqualsIgnoreCase(String,String)
+	 * @see OpenmrsUtil#folderContains(File, String)
+	 */
+	@Test
+	public void folderContains_shouldReturnTrueWhenFolderContainsFileWithGivenName() {
+		File folder = createFolderInUserDirectory("folder");
+		String separator = FileSystems.getDefault().getSeparator();
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: There is alot of code duplication of the above line. Can't we have it as a class constant?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 11]
```diff
@@ -62,6 +71,14 @@
 	private static GlobalProperty luhnGP = new GlobalProperty(
 	        OpenmrsConstants.GLOBAL_PROPERTY_DEFAULT_PATIENT_IDENTIFIER_VALIDATOR,
 	        OpenmrsConstants.LUHN_IDENTIFIER_VALIDATOR);
+
+	private File createFolderInUserDirectory(String name) {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Well according to what I have realized we are testing files in folders not folders in directories. Lets make this method to create a file given it's relative path name eg. 

```java
String relativePath = "{folderName}" + seperator + "{fileName}"; // The relative path could be the folder name + the file we wanna test
```

That way we can make code more reusable

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 12]
```diff
@@ -666,9 +683,169 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+		assertFalse(OpenmrsUtil.folderContains(null,"file"));
+	}
+	
+	/**
+	 * @see OpenmrsUtil#conceptListHelper(String)
+	 */
+	@Test
+	public void conceptListHelper_shouldReturnEmptyListWhenGivenDescriptorIsNullOrEmpty() {
+		List<Concept> list = OpenmrsUtil.conceptListHelper("");
+		assertTrue(list.size() == 0);
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Doesn't `List` as a Collection have a defined method to check whether it's empty or not?

Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Do we even need the variable! Isn't this just enough?
`assertTrue(OpenmrsUtil.conceptListHelper("").isEmpty());`

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 13]
```diff
@@ -666,9 +683,169 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+	}
+
+	/**
+	 * @see OpenmrsUtil#conceptListHelper(String)
+	 */
+	@Test
+	public void conceptListHelper_shouldReturnListOfConceptsWithGivenNameInDescriptor() {
+		List<Concept> helperList = OpenmrsUtil.conceptListHelper("name:YES|name:NO|name:DIED|set:25|set:26");
+		List<Concept> serviceList = new ArrayList<>();
+		List<Concept> aux;
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: What does `aux` mean? Can we use more descriptive variable names to self document your code?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 14]
```diff
@@ -666,9 +683,169 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+		Concept con;
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Same case to this one. More descriptive names needed

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 13 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I used it to retrieve the list of concepts by concept set . Is just an auxiliar to make code look better . I can get rid of it by adding the line directly as argument . Same with con ( which stands for concept).

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 15]
```diff
@@ -666,9 +683,169 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+		//Adding the concepts in the list to compare with the first one
+		ConceptService cs = Context.getConceptService();
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I see alot of code duplication of the above line. Can we have it as a class variable?

Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: ok

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 12 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Ok . I will do it like that .

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 10 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: ok

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 11 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: ok

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 9 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I can say something like catch (IOException e){Assert.fail();}

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 13 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: what I trying to say have more descriptive variable names. For `con` I understand it's `concept` but whats up with `concept`! But  if one names it like `drugConcept` it makes more sense!

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 9 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Is this of help https://stackoverflow.com/questions/16596418/how-to-handle-exceptions-in-junit/16597164#16597164?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 15 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: If i do that i get exception : no service found APIexception

**Comment by Andu033** [comment]
@samuelmale  did the changes. Unless the one with Concept service because it thrown exception . But i think that it s ok . I use this like of code 2 times so it wont save much space in test class. This was the final ajustment ? Is the ready to merge?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 15 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Make it a class level variable and call `Context.getConceptService();` from a method level. eg.

```java
@Before
public void setup() {
   // Initialization goes here
}
```

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 16]
```diff
@@ -63,6 +72,14 @@
 	        OpenmrsConstants.GLOBAL_PROPERTY_DEFAULT_PATIENT_IDENTIFIER_VALIDATOR,
 	        OpenmrsConstants.LUHN_IDENTIFIER_VALIDATOR);
 	
+	private final String separator = FileSystems.getDefault().getSeparator();
+	
+	private File createFileAtRelativePath(String relative_path) {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: For `relative_path`  and `current_path ` make them camel case sensitive.

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 17]
```diff
@@ -63,6 +72,14 @@
...
+		String current_path = System.getProperty("user.dir");
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: By the way can you use a better naming for `current_path` ?

**Comment by Andu033** [comment]
OK @samuelmale  . I pushed the changes . Anything else?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 18]
```diff
@@ -63,6 +72,19 @@
...
 	        OpenmrsConstants.LUHN_IDENTIFIER_VALIDATOR);
 	
+	private ConceptService cs;
+	private final String separator = FileSystems.getDefault().getSeparator();
+	
+	@Before
+	public void setup() {
+		cs=Context.getConceptService();
+	}
+	private File createFileAtRelativePath(String relativePath) {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: What do you mean by `createFileAtRelativePath` ? Doesn't `createFileInUserDirectory` sound better?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 19]
```diff
@@ -63,6 +72,19 @@
...
+		String userPath = System.getProperty("user.dir");
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Doesn't `userDirectoryPath` give more meaning?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 20]
```diff
@@ -63,6 +72,19 @@
...
+		File file = new File(userPath + separator + relativePath);
+		return file;
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: How sure are you that the file is really created? from where do you invoke `File#createNewFile()` ?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 21]
```diff
@@ -666,9 +688,156 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+		folder.mkdir();
+		File file = createFileAtRelativePath("folder" + separator + "file");
+		try {
+			file.createNewFile();
+			assertTrue(OpenmrsUtil.folderContains(folder, "file"));
+			file.delete();
+
+		} 
+		catch (IOException e) {Assert.fail("Unable to create file");}
+		folder.delete();
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: There is no test case(Behaviour assertions) once  IOException is thrown!

Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: If the exception is thrown the test will fail. Is it incorrect?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 20 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I create the file in the method of test. This way i can make both directories or file after called this method

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 21 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I think that testing this Exception is out of scope . If we are able to create file the test will do it s job . If not test fail.

**Comment by Andu033** [comment]
@samuelmale ping

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 22]
```diff
@@ -63,6 +72,19 @@
...
 	
+	private ConceptService cs;
+	private final String separator = FileSystems.getDefault().getSeparator();
+	
+	@Before
+	public void setup() {
+		cs=Context.getConceptService();
+	}
+	private File createFileInUserDirectory(String relativePath) {
+		String userDirectoryPath = System.getProperty("user.dir");
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: For running testcases, are you able to use this method to make a directory in the ${user} directory `OpenmrsUtil.getDirectoryInApplicationDataDirectory()` ?

Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: It works in a way that when running tests, files are created in temp dummy files and when running the webapplication , the files are created in the expected directory.

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 21 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Can you implement that?

Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: This is already implemented . Thats why i added this assert at catch. If the test go to the catch it will fail.

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 22 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: When im doing that createnewfile() throw expection

**Comment by samuelmale** [comment]
You have not addressed all the comments

**Comment by Andu033** [comment]
@samuelmale can't create new file in the directory returned by this method. It throw IOException

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 22 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I can't use this method. It throw exception when I want to create file.

**Comment by Andu033** [comment]
@samuelmale  can we finish this review?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 22 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I believe that method was created for that! You should tweak around code and make it work!

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 23]
```diff
@@ -666,9 +688,156 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+		catch (IOException e) {Assert.fail("Unable to create file");}
+		folder.delete();
+	}
+	
+	/**
+	 * @see OpenmrsUtil#folderContains(File, String)
 	 */
 	@Test
+	public void folderContains_shouldReturnFalseWhenFolderDoesNotContainFileWithGivenName() {
+		File folder = createFileInUserDirectory("folder");
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Was this(createFileInUserDirectory) method introduced for creating a file/folder or both?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 24]
```diff
@@ -666,9 +688,156 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+		assertFalse(OpenmrsUtil.folderContains(folder,"file"));
+		folder.delete();
+	}
+	
+	/**
+	 * @see OpenmrsUtil#folderContains(File, String)
+	 */
+	@Test
+	public void folderContains_shouldReturnFalseWhenGivenStringIsNull() {
+		File folder = createFileInUserDirectory("folder");
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: Was this(createFileInUserDirectory) method introduced for creating a file/folder or both?

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java (hunk 23 above)
Comment (Andu033) [reply api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: This method create file. If I want to create directors I just call mkdir() on this file. I can't modify this to do both because this would add to much complexity (one more argument, at least one if etc). So it's better to add a line of code when I want folder (I thinks I used it two times.

File: api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java [hunk 25]
```diff
@@ -666,9 +688,156 @@ public void shortenedStackTrace_shouldReturnNullIfStackTraceIsNull() {
...
+	 */
+	@Test
+	public void conceptSetHelper_shouldReturnEmptyListWhenGivenDescriptorIsNullOrEmpty() {
+		assertTrue(OpenmrsUtil.conceptSetHelper("").isEmpty());
+		assertTrue(OpenmrsUtil.conceptSetHelper(null).isEmpty());
+	}
+	
+		/**
+         * @see OpenmrsUtil#nullSafeEqualsIgnoreCase(String,String)
+         */
```
Comment (haripriya999) [inline api/src/test/java/org/openmrs/util/OpenmrsUtilTest.java]: I think it would look neater if  @see at line 837 started below the ending curly brace of conceptSetHelper_shouldReturnEmptyListWhenGivenDescriptorIsNullOrEmpty() :)
